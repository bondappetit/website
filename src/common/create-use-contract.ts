import { useCallback, useMemo } from 'react';
import type { AbiItem } from 'web3-utils';
import type { ContractOptions } from 'web3-eth-contract';
import networks from '@bondappetit/networks';

import { config } from 'src/config';
import { useNetworkConfig } from './use-network-config';
import { EthereumNetworkError } from './ethereum-network-error';
import { useLibrary } from './use-library';

type ContractParameters = {
  abi: AbiItem[] | AbiItem;
  address?: string;
  options?: ContractOptions;
};

type Callback = (network: Network) => ContractParameters;

export type Network = typeof networks[keyof typeof networks];

export const createUseContract = <T>(cb: Callback) => () => {
  const library = useLibrary();
  const networkConfig = useNetworkConfig();

  return useMemo(() => {
    try {
      const contractParams = cb(networkConfig);

      return (new library.eth.Contract(
        contractParams?.abi,
        contractParams?.address,
        contractParams?.options
      ) as unknown) as T;
    } catch {
      return null;
    }
  }, [library, networkConfig]);
};

export const useDynamicContract = <T>(
  contractParameters?: ContractParameters,
  currentChainId?: number
) => {
  const library = useLibrary(
    config.CHAIN_BINANCE_IDS.includes(Number(currentChainId))
  );

  return useCallback(
    (address?: string, abi?: AbiItem[] | AbiItem) => {
      const currentAbi = contractParameters?.abi ?? abi;

      let contract: null | T = null;

      try {
        if (!currentAbi) {
          throw new Error('Abi is required');
        }

        contract = (new library.eth.Contract(
          currentAbi,
          address
        ) as unknown) as T;
      } catch {
        throw new EthereumNetworkError();
      }

      return contract;
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [library, currentChainId]
  );
};
