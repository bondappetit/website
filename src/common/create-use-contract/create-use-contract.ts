import { useMemo } from 'react';
import Web3 from 'web3';
import { useWeb3React } from '@web3-react/core';
import type { AbiItem } from 'web3-utils';
import type { ContractOptions, Contract } from 'web3-eth-contract';
import networks from '@bondappetit/networks';

import { config } from 'src/config';
import { useNetworkConfig } from '../use-network-config';

type Options = {
  abi: AbiItem[] | AbiItem;
  address?: string;
  options?: ContractOptions;
};

type Callback = (network: Network) => Options;

const web3 = new Web3(
  config.IS_DEV
    ? Web3.givenProvider
    : new Web3.providers.HttpProvider(config.MAINNET_URL)
);

export type Network = typeof networks[keyof typeof networks];

const getContractOptions = (
  cb: Callback | Options,
  networkConfig?: Network
) => {
  if (!networkConfig && typeof cb === 'object') {
    return cb;
  }

  if (networkConfig && typeof cb === 'function') {
    return cb(networkConfig);
  }
};

export const createUseContract = <T = Contract>(
  cb: Callback | Options
) => () => {
  const { library } = useWeb3React<Web3>();
  const networkConfig = useNetworkConfig();
  const web3OrLib = library ?? web3;

  return useMemo(() => {
    const contractOptions = getContractOptions(cb, networkConfig);

    if (!web3OrLib || !contractOptions) return null;

    return (new web3OrLib.eth.Contract(
      contractOptions.abi,
      contractOptions.address,
      contractOptions.options
    ) as unknown) as T;
  }, [web3OrLib, networkConfig]);
};
