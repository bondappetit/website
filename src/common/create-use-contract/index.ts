export * from './create-use-contract';
