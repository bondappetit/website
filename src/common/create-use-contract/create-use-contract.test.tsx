import { renderHook } from '@testing-library/react-hooks';
import networks from '@bondappetit/networks';
import { Contract } from 'web3-eth-contract';

import { createUseContract } from './create-use-contract';

const contractOptions = {
  abi: networks.development.contracts.Investment.abi,
  address: networks.development.contracts.Investment.address
};

it('createUseContract should init hook', () => {
  const useContract = createUseContract(contractOptions);

  expect(useContract).toBeInstanceOf(Function);
});

describe('when using created use contract hook', () => {
  let useContract: () => Contract | null;

  beforeEach(() => {
    useContract = createUseContract(contractOptions);
  });

  it('should return contract', () => {
    const { rerender, result } = renderHook(() => useContract());

    rerender();

    expect(result.current).toMatchSnapshot();
  });
});
