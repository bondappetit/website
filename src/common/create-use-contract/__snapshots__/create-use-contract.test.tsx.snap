// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when using created use contract hook should return contract 1`] = `
Contract {
  "BatchRequest": [Function],
  "_address": "0x47a4e51510669697C23DffDD46F2D3268D372268",
  "_jsonInterface": Array [
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "_cumulative",
          "type": "address",
        },
        Object {
          "internalType": "address",
          "name": "_bond",
          "type": "address",
        },
        Object {
          "internalType": "uint256",
          "name": "_bondTokenLockDate",
          "type": "uint256",
        },
        Object {
          "internalType": "address",
          "name": "_uniswapRouter",
          "type": "address",
        },
      ],
      "payable": undefined,
      "stateMutability": "nonpayable",
      "type": "constructor",
    },
    Object {
      "anonymous": false,
      "constant": undefined,
      "inputs": Array [
        Object {
          "indexed": false,
          "internalType": "uint256",
          "name": "newPrice",
          "type": "uint256",
        },
      ],
      "name": "BondPriceChanged",
      "payable": undefined,
      "signature": "0x6b4bad8a5f845b5369ba4d524448526b53b7fd96eaa8003e8d766331dad9fd3d",
      "type": "event",
    },
    Object {
      "anonymous": false,
      "constant": undefined,
      "inputs": Array [
        Object {
          "indexed": false,
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
      ],
      "name": "InvestTokenAllowed",
      "payable": undefined,
      "signature": "0x25dacf8d076093c5272121a9d819f9002f65548533e715e76e04522fa5a59fb1",
      "type": "event",
    },
    Object {
      "anonymous": false,
      "constant": undefined,
      "inputs": Array [
        Object {
          "indexed": false,
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
      ],
      "name": "InvestTokenDenied",
      "payable": undefined,
      "signature": "0x1a50e5e1b18452e94911b877cd2967e8b348bbdaba6d4992dd27e4be3b907b68",
      "type": "event",
    },
    Object {
      "anonymous": false,
      "constant": undefined,
      "inputs": Array [
        Object {
          "indexed": false,
          "internalType": "address",
          "name": "investor",
          "type": "address",
        },
        Object {
          "indexed": false,
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
        Object {
          "indexed": false,
          "internalType": "uint256",
          "name": "amount",
          "type": "uint256",
        },
        Object {
          "indexed": false,
          "internalType": "uint256",
          "name": "reward",
          "type": "uint256",
        },
      ],
      "name": "Invested",
      "payable": undefined,
      "signature": "0x93638e6cf18b4f6a839de65df95898221f5d9f7dc54b87d4b31d40297643266e",
      "type": "event",
    },
    Object {
      "anonymous": false,
      "constant": undefined,
      "inputs": Array [
        Object {
          "indexed": true,
          "internalType": "address",
          "name": "previousOwner",
          "type": "address",
        },
        Object {
          "indexed": true,
          "internalType": "address",
          "name": "newOwner",
          "type": "address",
        },
      ],
      "name": "OwnershipTransferred",
      "payable": undefined,
      "signature": "0x8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0",
      "type": "event",
    },
    Object {
      "anonymous": false,
      "constant": undefined,
      "inputs": Array [
        Object {
          "indexed": false,
          "internalType": "address",
          "name": "recipient",
          "type": "address",
        },
        Object {
          "indexed": false,
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
        Object {
          "indexed": false,
          "internalType": "uint256",
          "name": "amount",
          "type": "uint256",
        },
      ],
      "name": "Withdrawal",
      "payable": undefined,
      "signature": "0x2717ead6b9200dd235aad468c9809ea400fe33ac69b5bfaa6d3e90fc922b6398",
      "type": "event",
    },
    Object {
      "constant": true,
      "inputs": Array [],
      "name": "bond",
      "outputs": Array [
        Object {
          "internalType": "contract Bond",
          "name": "",
          "type": "address",
        },
      ],
      "payable": undefined,
      "signature": "0x64c9ec6f",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": true,
      "inputs": Array [],
      "name": "bondPrice",
      "outputs": Array [
        Object {
          "internalType": "uint256",
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": undefined,
      "signature": "0xd7ccfb0b",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": true,
      "inputs": Array [],
      "name": "bondTokenLockDate",
      "outputs": Array [
        Object {
          "internalType": "uint256",
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": undefined,
      "signature": "0xb419b0bc",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": true,
      "inputs": Array [],
      "name": "cumulative",
      "outputs": Array [
        Object {
          "internalType": "contract ERC20",
          "name": "",
          "type": "address",
        },
      ],
      "payable": undefined,
      "signature": "0x7ae176ff",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": true,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "",
          "type": "address",
        },
      ],
      "name": "investmentTokens",
      "outputs": Array [
        Object {
          "internalType": "bool",
          "name": "",
          "type": "bool",
        },
      ],
      "payable": undefined,
      "signature": "0xe7b9cdc8",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": true,
      "inputs": Array [],
      "name": "owner",
      "outputs": Array [
        Object {
          "internalType": "address",
          "name": "",
          "type": "address",
        },
      ],
      "payable": undefined,
      "signature": "0x8da5cb5b",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [],
      "name": "renounceOwnership",
      "outputs": Array [],
      "payable": undefined,
      "signature": "0x715018a6",
      "stateMutability": "nonpayable",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "newOwner",
          "type": "address",
        },
      ],
      "name": "transferOwnership",
      "outputs": Array [],
      "payable": undefined,
      "signature": "0xf2fde38b",
      "stateMutability": "nonpayable",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
      ],
      "name": "allowToken",
      "outputs": Array [],
      "payable": undefined,
      "signature": "0xb53472ef",
      "stateMutability": "nonpayable",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
      ],
      "name": "denyToken",
      "outputs": Array [],
      "payable": undefined,
      "signature": "0xc5ea3d83",
      "stateMutability": "nonpayable",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "uint256",
          "name": "newPrice",
          "type": "uint256",
        },
      ],
      "name": "changeBondPrice",
      "outputs": Array [],
      "payable": undefined,
      "signature": "0xab857038",
      "stateMutability": "nonpayable",
      "type": "function",
    },
    Object {
      "constant": true,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
        Object {
          "internalType": "uint256",
          "name": "amount",
          "type": "uint256",
        },
      ],
      "name": "price",
      "outputs": Array [
        Object {
          "internalType": "uint256",
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": undefined,
      "signature": "0xad9b8024",
      "stateMutability": "view",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "token",
          "type": "address",
        },
        Object {
          "internalType": "uint256",
          "name": "amount",
          "type": "uint256",
        },
      ],
      "name": "invest",
      "outputs": Array [
        Object {
          "internalType": "bool",
          "name": "",
          "type": "bool",
        },
      ],
      "payable": undefined,
      "signature": "0xb9b8c246",
      "stateMutability": "nonpayable",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [],
      "name": "investETH",
      "outputs": Array [
        Object {
          "internalType": "bool",
          "name": "",
          "type": "bool",
        },
      ],
      "payable": true,
      "signature": "0xa39cfae6",
      "stateMutability": "payable",
      "type": "function",
    },
    Object {
      "constant": undefined,
      "inputs": Array [
        Object {
          "internalType": "address",
          "name": "recipient",
          "type": "address",
        },
      ],
      "name": "withdraw",
      "outputs": Array [],
      "payable": undefined,
      "signature": "0x51cff8d9",
      "stateMutability": "nonpayable",
      "type": "function",
    },
  ],
  "_provider": HttpProvider {
    "agent": undefined,
    "connected": false,
    "headers": undefined,
    "host": "https://mainnet.infura.io/v3/undefined",
    "httpsAgent": Agent {
      "_events": Object {
        "free": [Function],
        "newListener": [Function],
      },
      "_eventsCount": 2,
      "_maxListeners": undefined,
      "_sessionCache": Object {
        "list": Array [],
        "map": Object {},
      },
      "defaultPort": 443,
      "freeSockets": Object {},
      "keepAlive": true,
      "keepAliveMsecs": 1000,
      "maxCachedSessions": 100,
      "maxFreeSockets": 256,
      "maxSockets": Infinity,
      "options": Object {
        "keepAlive": true,
        "path": null,
      },
      "protocol": "https:",
      "requests": Object {},
      "sockets": Object {},
      Symbol(kCapture): false,
    },
    "timeout": 0,
    "withCredentials": false,
  },
  "_requestManager": RequestManager {
    "provider": HttpProvider {
      "agent": undefined,
      "connected": false,
      "headers": undefined,
      "host": "https://mainnet.infura.io/v3/undefined",
      "httpsAgent": Agent {
        "_events": Object {
          "free": [Function],
          "newListener": [Function],
        },
        "_eventsCount": 2,
        "_maxListeners": undefined,
        "_sessionCache": Object {
          "list": Array [],
          "map": Object {},
        },
        "defaultPort": 443,
        "freeSockets": Object {},
        "keepAlive": true,
        "keepAliveMsecs": 1000,
        "maxCachedSessions": 100,
        "maxFreeSockets": 256,
        "maxSockets": Infinity,
        "options": Object {
          "keepAlive": true,
          "path": null,
        },
        "protocol": "https:",
        "requests": Object {},
        "sockets": Object {},
        Symbol(kCapture): false,
      },
      "timeout": 0,
      "withCredentials": false,
    },
    "providers": Object {
      "HttpProvider": [Function],
      "IpcProvider": [Function],
      "WebsocketProvider": [Function],
    },
    "subscriptions": Map {},
  },
  "clearSubscriptions": [Function],
  "currentProvider": HttpProvider {
    "agent": undefined,
    "connected": false,
    "headers": undefined,
    "host": "https://mainnet.infura.io/v3/undefined",
    "httpsAgent": Agent {
      "_events": Object {
        "free": [Function],
        "newListener": [Function],
      },
      "_eventsCount": 2,
      "_maxListeners": undefined,
      "_sessionCache": Object {
        "list": Array [],
        "map": Object {},
      },
      "defaultPort": 443,
      "freeSockets": Object {},
      "keepAlive": true,
      "keepAliveMsecs": 1000,
      "maxCachedSessions": 100,
      "maxFreeSockets": 256,
      "maxSockets": Infinity,
      "options": Object {
        "keepAlive": true,
        "path": null,
      },
      "protocol": "https:",
      "requests": Object {},
      "sockets": Object {},
      Symbol(kCapture): false,
    },
    "timeout": 0,
    "withCredentials": false,
  },
  "defaultAccount": null,
  "defaultBlock": "latest",
  "defaultChain": undefined,
  "defaultCommon": undefined,
  "defaultHardfork": undefined,
  "events": Object {
    "0x1a50e5e1b18452e94911b877cd2967e8b348bbdaba6d4992dd27e4be3b907b68": [Function],
    "0x25dacf8d076093c5272121a9d819f9002f65548533e715e76e04522fa5a59fb1": [Function],
    "0x2717ead6b9200dd235aad468c9809ea400fe33ac69b5bfaa6d3e90fc922b6398": [Function],
    "0x6b4bad8a5f845b5369ba4d524448526b53b7fd96eaa8003e8d766331dad9fd3d": [Function],
    "0x8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0": [Function],
    "0x93638e6cf18b4f6a839de65df95898221f5d9f7dc54b87d4b31d40297643266e": [Function],
    "BondPriceChanged": [Function],
    "BondPriceChanged(uint256)": [Function],
    "InvestTokenAllowed": [Function],
    "InvestTokenAllowed(address)": [Function],
    "InvestTokenDenied": [Function],
    "InvestTokenDenied(address)": [Function],
    "Invested": [Function],
    "Invested(address,address,uint256,uint256)": [Function],
    "OwnershipTransferred": [Function],
    "OwnershipTransferred(address,address)": [Function],
    "Withdrawal": [Function],
    "Withdrawal(address,address,uint256)": [Function],
    "allEvents": [Function],
  },
  "extend": [Function],
  "givenProvider": null,
  "handleRevert": false,
  "methods": Object {
    "0x51cff8d9": [Function],
    "0x64c9ec6f": [Function],
    "0x715018a6": [Function],
    "0x7ae176ff": [Function],
    "0x8da5cb5b": [Function],
    "0xa39cfae6": [Function],
    "0xab857038": [Function],
    "0xad9b8024": [Function],
    "0xb419b0bc": [Function],
    "0xb53472ef": [Function],
    "0xb9b8c246": [Function],
    "0xc5ea3d83": [Function],
    "0xd7ccfb0b": [Function],
    "0xe7b9cdc8": [Function],
    "0xf2fde38b": [Function],
    "allowToken": [Function],
    "allowToken(address)": [Function],
    "bond": [Function],
    "bond()": [Function],
    "bondPrice": [Function],
    "bondPrice()": [Function],
    "bondTokenLockDate": [Function],
    "bondTokenLockDate()": [Function],
    "changeBondPrice": [Function],
    "changeBondPrice(uint256)": [Function],
    "cumulative": [Function],
    "cumulative()": [Function],
    "denyToken": [Function],
    "denyToken(address)": [Function],
    "invest": [Function],
    "invest(address,uint256)": [Function],
    "investETH": [Function],
    "investETH()": [Function],
    "investmentTokens": [Function],
    "investmentTokens(address)": [Function],
    "owner": [Function],
    "owner()": [Function],
    "price": [Function],
    "price(address,uint256)": [Function],
    "renounceOwnership": [Function],
    "renounceOwnership()": [Function],
    "transferOwnership": [Function],
    "transferOwnership(address)": [Function],
    "withdraw": [Function],
    "withdraw(address)": [Function],
  },
  "options": Object {
    "address": "0x47a4e51510669697C23DffDD46F2D3268D372268",
    "jsonInterface": Array [
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "_cumulative",
            "type": "address",
          },
          Object {
            "internalType": "address",
            "name": "_bond",
            "type": "address",
          },
          Object {
            "internalType": "uint256",
            "name": "_bondTokenLockDate",
            "type": "uint256",
          },
          Object {
            "internalType": "address",
            "name": "_uniswapRouter",
            "type": "address",
          },
        ],
        "payable": undefined,
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      Object {
        "anonymous": false,
        "constant": undefined,
        "inputs": Array [
          Object {
            "indexed": false,
            "internalType": "uint256",
            "name": "newPrice",
            "type": "uint256",
          },
        ],
        "name": "BondPriceChanged",
        "payable": undefined,
        "signature": "0x6b4bad8a5f845b5369ba4d524448526b53b7fd96eaa8003e8d766331dad9fd3d",
        "type": "event",
      },
      Object {
        "anonymous": false,
        "constant": undefined,
        "inputs": Array [
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
        ],
        "name": "InvestTokenAllowed",
        "payable": undefined,
        "signature": "0x25dacf8d076093c5272121a9d819f9002f65548533e715e76e04522fa5a59fb1",
        "type": "event",
      },
      Object {
        "anonymous": false,
        "constant": undefined,
        "inputs": Array [
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
        ],
        "name": "InvestTokenDenied",
        "payable": undefined,
        "signature": "0x1a50e5e1b18452e94911b877cd2967e8b348bbdaba6d4992dd27e4be3b907b68",
        "type": "event",
      },
      Object {
        "anonymous": false,
        "constant": undefined,
        "inputs": Array [
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "investor",
            "type": "address",
          },
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
          Object {
            "indexed": false,
            "internalType": "uint256",
            "name": "amount",
            "type": "uint256",
          },
          Object {
            "indexed": false,
            "internalType": "uint256",
            "name": "reward",
            "type": "uint256",
          },
        ],
        "name": "Invested",
        "payable": undefined,
        "signature": "0x93638e6cf18b4f6a839de65df95898221f5d9f7dc54b87d4b31d40297643266e",
        "type": "event",
      },
      Object {
        "anonymous": false,
        "constant": undefined,
        "inputs": Array [
          Object {
            "indexed": true,
            "internalType": "address",
            "name": "previousOwner",
            "type": "address",
          },
          Object {
            "indexed": true,
            "internalType": "address",
            "name": "newOwner",
            "type": "address",
          },
        ],
        "name": "OwnershipTransferred",
        "payable": undefined,
        "signature": "0x8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0",
        "type": "event",
      },
      Object {
        "anonymous": false,
        "constant": undefined,
        "inputs": Array [
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "recipient",
            "type": "address",
          },
          Object {
            "indexed": false,
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
          Object {
            "indexed": false,
            "internalType": "uint256",
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "Withdrawal",
        "payable": undefined,
        "signature": "0x2717ead6b9200dd235aad468c9809ea400fe33ac69b5bfaa6d3e90fc922b6398",
        "type": "event",
      },
      Object {
        "constant": true,
        "inputs": Array [],
        "name": "bond",
        "outputs": Array [
          Object {
            "internalType": "contract Bond",
            "name": "",
            "type": "address",
          },
        ],
        "payable": undefined,
        "signature": "0x64c9ec6f",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": true,
        "inputs": Array [],
        "name": "bondPrice",
        "outputs": Array [
          Object {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "payable": undefined,
        "signature": "0xd7ccfb0b",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": true,
        "inputs": Array [],
        "name": "bondTokenLockDate",
        "outputs": Array [
          Object {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "payable": undefined,
        "signature": "0xb419b0bc",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": true,
        "inputs": Array [],
        "name": "cumulative",
        "outputs": Array [
          Object {
            "internalType": "contract ERC20",
            "name": "",
            "type": "address",
          },
        ],
        "payable": undefined,
        "signature": "0x7ae176ff",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": true,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "name": "investmentTokens",
        "outputs": Array [
          Object {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "payable": undefined,
        "signature": "0xe7b9cdc8",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": true,
        "inputs": Array [],
        "name": "owner",
        "outputs": Array [
          Object {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "payable": undefined,
        "signature": "0x8da5cb5b",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [],
        "name": "renounceOwnership",
        "outputs": Array [],
        "payable": undefined,
        "signature": "0x715018a6",
        "stateMutability": "nonpayable",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "newOwner",
            "type": "address",
          },
        ],
        "name": "transferOwnership",
        "outputs": Array [],
        "payable": undefined,
        "signature": "0xf2fde38b",
        "stateMutability": "nonpayable",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
        ],
        "name": "allowToken",
        "outputs": Array [],
        "payable": undefined,
        "signature": "0xb53472ef",
        "stateMutability": "nonpayable",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
        ],
        "name": "denyToken",
        "outputs": Array [],
        "payable": undefined,
        "signature": "0xc5ea3d83",
        "stateMutability": "nonpayable",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "uint256",
            "name": "newPrice",
            "type": "uint256",
          },
        ],
        "name": "changeBondPrice",
        "outputs": Array [],
        "payable": undefined,
        "signature": "0xab857038",
        "stateMutability": "nonpayable",
        "type": "function",
      },
      Object {
        "constant": true,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
          Object {
            "internalType": "uint256",
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "price",
        "outputs": Array [
          Object {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "payable": undefined,
        "signature": "0xad9b8024",
        "stateMutability": "view",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "token",
            "type": "address",
          },
          Object {
            "internalType": "uint256",
            "name": "amount",
            "type": "uint256",
          },
        ],
        "name": "invest",
        "outputs": Array [
          Object {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "payable": undefined,
        "signature": "0xb9b8c246",
        "stateMutability": "nonpayable",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [],
        "name": "investETH",
        "outputs": Array [
          Object {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "payable": true,
        "signature": "0xa39cfae6",
        "stateMutability": "payable",
        "type": "function",
      },
      Object {
        "constant": undefined,
        "inputs": Array [
          Object {
            "internalType": "address",
            "name": "recipient",
            "type": "address",
          },
        ],
        "name": "withdraw",
        "outputs": Array [],
        "payable": undefined,
        "signature": "0x51cff8d9",
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
  },
  "providers": Object {
    "HttpProvider": [Function],
    "IpcProvider": [Function],
    "WebsocketProvider": [Function],
  },
  "setProvider": [Function],
  "setRequestManager": [Function],
  "transactionBlockTimeout": 50,
  "transactionConfirmationBlocks": 24,
  "transactionPollingTimeout": 750,
}
`;
